simple_learn/classifiers/simple_classifier.py

```diff
@@ -127,41 +128,50 @@ def fit(self, train_x, train_y, folds=3):
         folds : int, optional
             The number of folds for cross validation
         """
-
-        estimators = all_estimators(type_filter="classifier")
-        for name, ClassifierClass in estimators:
-            if name in model_param_map:
-                param_grid = model_param_map[name]
-                grid_clf = GridSearchCV(
-                    ClassifierClass(),
-                    param_grid,
-                    cv=folds,
-                    scoring="accuracy",
-                    verbose=0,
-                    n_jobs=-1,
-                )
-                start = time.time()
-                try:
-                    grid_clf.fit(train_x, train_y)
-                except BaseException as error:
-                    self.failed_models.append(name)
-                    self.logger.warning(f"{name} failed due to, Error : {error}.")
-                    continue
-                end = time.time()
-                if grid_clf.best_score_ > self.metrics.get("Training Accuracy", 0.0):
-                    self.metrics["Training Accuracy"] = grid_clf.best_score_
-                    pred_y = grid_clf.predict(train_x)
-                    self.metrics["Jaccard Score"] = jaccard_score(
-                        train_y, pred_y, average="macro"
-                    )
-                    self.metrics["F1 Score"] = f1_score(
-                        train_y, pred_y, average="macro"
+        with tqdm(
+            total=(len(model_param_map)),
+            desc="Fitting Models",
+            unit=" Algorithm",
+            ncols=100,
+        ) as progressbar:
+            estimators = all_estimators(type_filter="classifier")
+            for name, ClassifierClass in estimators:
+                if name in model_param_map:
+                    param_grid = model_param_map[name]
+                    grid_clf = GridSearchCV(
+                        ClassifierClass(),
+                        param_grid,
+                        cv=folds,
+                        scoring="accuracy",
+                        verbose=0,
+                        n_jobs=-1,
                     )
-                    self.sk_model = grid_clf.best_estimator_
-                    self.name = name
-                    self.attributes = grid_clf.best_params_
-                    self.train_duration = grid_clf.refit_time_
-                    self.gridsearch_duration = end - start
+                    progressbar.update(1)
+                    start = time.time()
+                    try:
+                        grid_clf.fit(train_x, train_y)
+                    except BaseException as error:
+                        self.failed_models.append(name)
```

Comment:
Since the error logging is getting in the way of the progress bar, let's go ahead and change this to log the errors after fitting:
```python
 try:
    grid_clf.fit(train_x, train_y)
except BaseException as error:
    self.failed_models.append((name, error))
    continue
# After fit
for algorithm, error in self.failed_models:
    self.logger.warning(f"{algorithm} failed due to, Error : {error}.")
```

Reply:
Unfortunately this is the result now. I think it might be better to just maintain the original so only one error is reported?

```
Fitting Models:  43%|█████████████████▌                       | 3/7 [00:03<00:05,  1.33s/ Algorithm]DecisionTreeRegressor failed due to, Error : 'poisson'.
Fitting Models:  57%|███████████████████████▍                 | 4/7 [00:48<00:55, 18.35s/ Algorithm]DecisionTreeRegressor failed due to, Error : 'poisson'.
Fitting Models:  71%|█████████████████████████████▎           | 5/7 [00:51<00:25, 12.78s/ Algorithm]DecisionTreeRegressor failed due to, Error : 'poisson'.
Fitting Models:  86%|███████████████████████████████████▏     | 6/7 [00:51<00:08,  8.55s/ Algorithm]DecisionTreeRegressor failed due to, Error : 'poisson'.
Fitting Models: 100%|█████████████████████████████████████████| 7/7 [01:07<00:00, 10.85s/ Algorithm]DecisionTreeRegressor failed due to, Error : 'poisson'.
Fitting Models: 100%|█████████████████████████████████████████| 7/7 [01:08<00:00,  9.74s/ Algorithm]
```

Reply:
That's interesting, did you add the logging functions outside the `tqdm` loop? Feel free to push your changes so we can go ahead and merge, I can look to debug it. I think it has something to do with where the logging function is located.

Reply:
I've set the errors as a progress description so it doesn't reload the bar after every error. I'm not sure if this is better but I gave it bash.

```python
try:
      grid_rgr.fit(train_x, train_y)
except BaseException as error:
      self.failed_models.append((name, error))
      continue
# After fit
for algorithm, error in self.failed_models:
      progressbar.set_description(f"{algorithm} failed due to, Error : {error}.")
      # self.logger.warning(f"{algorithm} failed due to, Error : {error}.")
```

```
>>> from sklearn.datasets import load_boston
>>> from simple_learn.regressors import SimpleRegressor
>>> boston = load_boston() 
>>> true_x = boston.data   
>>> true_y = boston.target 
>>> clf = SimpleRegressor()
>>> clf.fit(true_x, true_y)
DecisionTreeRegressor failed due to, Error : 'poisson'.: 100%|██████████████████████████████████████████████████| 7/7 [01:08<00:00,  9.73s/ Algorithm]
>>>
```

Reply:
I see, what does the output look like if you log the failed models after the `tqdm` loop.

Reply:
Tqdm isn't looping here right? It just sits on top of the loop and the progressbar.update(1) just adds 1 to the bar after each loop. I found a solution for the logging disrupting the bar though. Found a custom logging solution. If you want to tackle it. I can do the push and revert all of this, I'm going to be slightly scarce due to interviews and class for the next few days. 

**custom logging module**
```python
import logging
import tqdm

class TqdmLoggingHandler(logging.Handler):
    def __init__(self, level=logging.NOTSET):
        super().__init__(level)

    def emit(self, record):
        try:
            msg = self.format(record)
            tqdm.tqdm.write(msg)
            self.flush()
        except (KeyboardInterrupt, SystemExit):
            raise
        except:
            self.handleError(record)  
```

**Code used in the simple regressor**
```python
# Custom logging
log = logging.getLogger(__name__)
log.setLevel(logging.INFO)
log.addHandler(custom_logging.TqdmLoggingHandler())
 with tqdm(.....)

try:
    grid_rgr.fit(train_x, train_y)
except BaseException as error:
      self.failed_models.append(name)
      log.info(f"{name} failed due to, Error : {error}.")
     continue
```
**Output**
```
>>> clf.fit(true_x, true_y)
DecisionTreeRegressor failed due to, Error : 'poisson'.
Fitting Models: 100%|███████████████████████████████████████████████████████████████████████████████████████████| 7/7 [01:05<00:00,  9.37s/ Algorithm]
```

Reply:
Awesome, this looks good 👍 . So does the bar continue on the next line when when a failure is logged?

Reply:
@skekre98 yes the bar doesn't duplicate because of each error now and the error appears above the bar.

Reply:
Awesome, can you push these changes. They look good to merge :)